### DISCUSSION TITLE
Custom Loss Function: Integrating Boundary Loss with YOLOv8 for Segmentation

### DISCUSSION THREAD

**Discussion by MarioPasc** [ultralytics/ultralytics#12844] [Q&A]
# Context
I am working on a research project focused on segmenting small lesions caused by Epilepsy FCD type II. The lesions are small, creating a significant imbalance between the lesion area and the normal area, making segmentation challenging. To improve segmentation performance, I aim to integrate a boundary loss function proposed in a medical imaging paper with YOLOv8 (https://github.com/LIVIAETS/boundary-loss). 

My objective is to combine the existing loss functions in YOLOv8 with a boundary loss specifically designed to enhance the detection of small objects by focusing on the boundaries. This custom boundary loss is calculated as a pixel-wise multiplication between the network predictions (softmax outputs) and a pre-computed distance map. I think integration should help in segmenting small lesions more accurately.

This boundary loss, as mentioned in their github repository, involves three key components:
1. **Boundary Loss Function**: This computes the loss using network predictions and pre-computed distance maps.
2. **Distance Map Function**: Generates distance maps from the ground truth masks.
3. **Dataloader Transforms**: Prepares the distance maps during data loading.

Here is a brief overview of the boundary loss function I'm planning to use:

```python
class SurfaceLoss():
    def __init__(self, **kwargs):
        self.idc: List[int] = kwargs["idc"]

    def __call__(self, probs: Tensor, dist_maps: Tensor) -> Tensor:
        assert simplex(probs)
        assert not one_hot(dist_maps)

        pc = probs[:, self.idc, ...].type(torch.float32)
        dc = dist_maps[:, self.idc, ...].type(torch.float32)

        multipled = einsum("bkwh,bkwh->bkwh", pc, dc)
        loss = multipled.mean()
        return loss

BoundaryLoss = SurfaceLoss
```
# Current implementation
This is the idea I had, take into consideration that the task is segmentation. It would be really helpful if you could take a look at it:

```python
import torch
import torch.nn.functional as F
from ultralytics.yolo.utils.loss import v8DetectionLoss
from utils.boundary_loss import BoundaryLoss     # Custom loss
from ultralytics.yolo.utils.ops import make_anchors, xyxy2xywh

class v8SegmentationLoss(v8DetectionLoss):
    def __init__(self, model):
        super().__init__(model)
        self.nm = model.model[-1].nm  # number of masks
        self.overlap = model.args.overlap_mask
        self.boundary_loss = BoundaryLoss(idc=[1])  # Assuming the target class has index 1

    def __call__(self, preds, batch):
        loss = torch.zeros(5, device=self.device)  # box, cls, dfl, boundary
        feats, pred_masks, proto = preds if len(preds) == 3 else preds[1]
        batch_size, _, mask_h, mask_w = proto.shape
        pred_distri, pred_scores = torch.cat([xi.view(feats[0].shape[0], self.no, -1) for xi in feats], 2).split(
            (self.reg_max * 4, self.nc), 1)

        pred_scores = pred_scores.permute(0, 2, 1).contiguous()
        pred_distri = pred_distri.permute(0, 2, 1).contiguous()
        pred_masks = pred_masks.permute(0, 2, 1).contiguous()

        dtype = pred_scores.dtype
        imgsz = torch.tensor(feats[0].shape[2:], device=self.device, dtype=dtype) * self.stride[0]
        anchor_points, stride_tensor = make_anchors(feats, self.stride, 0.5)

        try:
            batch_idx = batch['batch_idx'].view(-1, 1)
            targets = torch.cat((batch_idx, batch['cls'].view(-1, 1), batch['bboxes']), 1)
            targets = self.preprocess(targets.to(self.device), batch_size, scale_tensor=imgsz[[1, 0, 1, 0]])
            gt_labels, gt_bboxes = targets.split((1, 4), 2)
            mask_gt = gt_bboxes.sum(2, keepdim=True).gt_(0)
        except RuntimeError as e:
            raise TypeError('ERROR ❌ segment dataset incorrectly formatted or not a segment dataset.\n'
                            "This error can occur when incorrectly training a 'segment' model on a 'detect' dataset, "
                            "i.e. 'yolo train model=yolov8n-seg.pt data=coco128.yaml'.\nVerify your dataset is a "
                            "correctly formatted 'segment' dataset using 'data=coco128-seg.yaml' "
                            'as an example.\nSee https://docs.ultralytics.com/tasks/segment/ for help.') from e

        pred_bboxes = self.bbox_decode(anchor_points, pred_distri)

        _, target_bboxes, target_scores, fg_mask, target_gt_idx = self.assigner(
            pred_scores.detach().sigmoid(), (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
            anchor_points * stride_tensor, gt_labels, gt_bboxes, mask_gt)

        target_scores_sum = max(target_scores.sum(), 1)

        loss[2] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum

        if fg_mask.sum():
            loss[0], loss[3] = self.bbox_loss(pred_distri, pred_bboxes, anchor_points, target_bboxes / stride_tensor,
                                              target_scores, target_scores_sum, fg_mask)
            masks = batch['masks'].to(self.device).float()
            dist_maps = batch['dist_maps'].to(self.device).float()

            if tuple(masks.shape[-2:]) != (mask_h, mask_w):
                masks = F.interpolate(masks[None], (mask_h, mask_w), mode='nearest')[0]
                dist_maps = F.interpolate(dist_maps[None], (mask_h, mask_w), mode='nearest')[0]

            for i in range(batch_size):
                if fg_mask[i].sum():
                    mask_idx = target_gt_idx[i][fg_mask[i]]
                    if self.overlap:
                        gt_mask = torch.where(masks[[i]] == (mask_idx + 1).view(-1, 1, 1), 1.0, 0.0)
                    else:
                        gt_mask = masks[batch_idx.view(-1) == i][mask_idx]
                    xyxyn = target_bboxes[i][fg_mask[i]] / imgsz[[1, 0, 1, 0]]
                    marea = xyxy2xywh(xyxyn)[:, 2:].prod(1)
                    mxyxy = xyxyn * torch.tensor([mask_w, mask_h, mask_w, mask_h], device=self.device)
                    loss[1] += self.single_mask_loss(gt_mask, pred_masks[i][fg_mask[i]], proto[i], mxyxy, marea)

                    dist_map = dist_maps[batch_idx.view(-1) == i][mask_idx]
                    probs = torch.softmax(pred_masks[i][fg_mask[i]] @ proto[i], dim=1)
                    loss[4] += self.boundary_loss(probs, dist_map)

                else:
                    loss[1] += (proto * 0).sum() + (pred_masks * 0).sum()
                    loss[4] += (dist_maps * 0).sum()

        else:
            loss[1] += (proto * 0).sum() + (pred_masks * 0).sum()
            loss[4] += (dist_maps * 0).sum()

        loss[0] *= self.hyp.box
        loss[1] *= self.hyp.box / batch_size
        loss[2] *= self.hyp.cls
        loss[3] *= self.hyp.dfl
        loss[4] *= self.hyp.boundary

        return loss.sum() * batch_size, loss.detach()
```
# Questions
- What is the best way to integrate the boundary loss with the existing loss functions in YOLOv8? 
- Are there any recommended practices or steps to verify that the new loss function is correctly integrated and functioning as expected? 
- Are there any potential issues or considerations we should be aware of when combining the boundary loss with YOLOv8?

I appreciate your guidance and any insights you can provide to help me effectively integrate this loss function into YOLOv8.

**Comment by glenn-jocher** [comment]
@MarioPasc hello! It's great to see you're working on enhancing YOLOv8 for segmenting small lesions with a custom boundary loss. Here’s how you can integrate and verify your custom loss function:

1. **Integration**: You've already made a good start by defining your `v8SegmentationLoss` class that inherits from `v8DetectionLoss`. To integrate the boundary loss, ensure that your custom loss class correctly handles the inputs and outputs as you've outlined. It looks like you're on the right track by adding the boundary loss computation within the `__call__` method of your custom loss class.

2. **Verification**:
   - **Unit Testing**: Create small, manageable tests that check the boundary loss function with known inputs and outputs.
   - **Debug Prints**: Temporarily add print statements in your loss function to ensure that tensors are of expected shapes and values.
   - **Visual Inspection**: After training for a few epochs, visually inspect the segmentation results to see if the model is focusing more accurately on the boundaries.

3. **Potential Issues**:
   - **Gradient Flow**: Ensure that the gradients are flowing correctly through your custom loss components. You might want to check if the gradients are vanishing or exploding, especially since boundary loss involves operations that could potentially lead to unstable gradients.
   - **Balance of Loss Components**: The scale of the boundary loss might need adjustment to ensure it doesn’t overpower other loss components. Consider using a weighting factor to balance its contribution to the total loss.

Here’s a small snippet on how you might adjust the weighting of your boundary loss:

```python
class v8SegmentationLoss(v8DetectionLoss):
    def __init__(self, model, boundary_loss_weight=1.0):
        super().__init__(model)
        self.boundary_loss_weight = boundary_loss_weight
        # other initializations

    def __call__(self, preds, batch):
        # existing code
        boundary_loss = self.boundary_loss(probs, dist_map)
        loss[4] += boundary_loss * self.boundary_loss_weight
        # existing code
        return loss.sum() * batch_size, loss.detach()
```

Adjust `boundary_loss_weight` to tune the contribution of the boundary loss to the total loss.

Feel free to reach out if you encounter specific issues or need further assistance. Good luck with your project! 🚀
